### ISSUE TITLE
Should rejection be cumulative between `FittingWithOutlierRemoval` iterations?

### ISSUE THREAD

**Issue by jehturner** [astropy/astropy#7424]
See the discussion in #7350 (which doesn't really belong there).

In summary, currently, the rejection of points is cumulative from one iteration to the next. It's possible that some points will appear deviant at an early iteration, due to the fit not being accurate yet, but then turn out to fall closer to the final fit. So should the algorithm in fact restore those points between iterations and do the rejection from scratch? Perhaps the answer is that cumulative rejection converges correctly because of the way the standard deviation scales down between iterations?

It turns out that cumulative masking behaves reasonably and is the same thing IRAF fitting does, so I think that's a good baseline, but @ysBach has argued for non-cumulative rejection.

**Comment by mcara** [comment]
I am not sure I understood the entire issue. What I understood is that in step2 in https://github.com/astropy/astropy/issues/7350#issuecomment-381054895, pixels that were clipped are added to the mask and not removed from it in the next iteration and therefore, instead of reseting the mask at each iteration you keep accumulating "bad" points. If this is a correct description of the issue, then I agree with @ysBach - mask should be reset before each iteration.

**Comment by nden** [comment]
This issue was discussed at PyAstro2019 and the consensus was that both ways of fitting are needed. Adding a flag controlling the behavior may be one way to do it.

**Comment by ysBach** [comment]
While I was testing my own sigma-clipping code against IRAF and astropy, I found few examples where *cumulative* rejection does raise problem. Although it does not explicitly use Fitting, mathematically this is identical to "fitting a constant function to data", so I guess it is relevant to this issue.


```python
import numpy as np
from astropy.stats import sigma_clip
%config InlineBackend.figure_format = 'retina'
from matplotlib import pyplot as plt

np.random.seed(12345)
ix, iy = 6, 5
data3d = np.random.normal(size=(7, 10, 10))
data1d = np.sort(data3d[:, ix, iy])  # Sort for better readability


print("data       : ", data1d)

fig, axs = plt.subplots(1, 1, figsize=(8, 5))
ax = axs
cens, lows, upps = [np.median(data1d)], [], []
maxiters = np.arange(1, 4)
for it in maxiters:
    print(f"maxiters {it}")
    sc1d = sigma_clip(data1d, sigma=1, cenfunc='median', maxiters=it, axis=0, return_bounds=True)
    cens.append(np.ma.median(sc1d[0]))
    lows.append(sc1d[1])
    upps.append(sc1d[2])
    print("mask: ", sc1d[0].mask, "bounds: ", sc1d[1], sc1d[2])
    
    ndata = data1d.size
    nmask = np.count_nonzero(sc1d[0].mask)
    ax.plot(np.repeat(it, nmask), data1d[sc1d[0].mask], 'kx')
    ax.plot(np.repeat(it, ndata - nmask), data1d[~sc1d[0].mask], 'k.')
    
ax.plot(maxiters, cens[:-1], 'b:', lw=1, label="Center of rejection")  
ax.plot(maxiters, cens[1:], 'g-.', lw=1, label="Center after sigma-clipping")
ax.plot(maxiters, lows, 'r-', lw=1, label="Bounds")
ax.plot(maxiters, upps, 'r-', lw=1)
ax.legend(bbox_to_anchor=(1., 1.))
ax.set(xlabel="maxiters", ylabel="value")
plt.tight_layout()
plt.show()

# plt.savefig(figoutpath, dpi=300, bbox_inches = "tight")
```
Note when maxiters = 2, the **6-th element is masked** even though the value (0.746...) is within the bounds (0.132..., 0.767...):
```
data       :  [-1.4373956  -1.28175713 -1.21830198 -0.15675953  0.44988042  0.56710603 0.74643304]
maxiters 1
mask:  [ True  True  True False False False  True] bounds:  -1.0450096836363323 0.7314906193328572
maxiters 2
mask:  [ True  True  True  True False False  True] bounds:  0.13264667354834941 0.7671141567593287
maxiters 3
mask:  [ True  True  True  True False False  True] bounds:  0.449880415153839 0.5671060292895793
```

![image](https://user-images.githubusercontent.com/17214293/88004366-d845fe80-cb41-11ea-8641-42f76337ecbd.png)



Similar thing happens when ``ix, iy = 6, 5``:
```
data       :  [-0.97793623 -0.95586885 -0.57430599  0.10643131  1.30642518  1.44016687  1.66201449]
maxiters 1
mask:  [ True False False False  True  True  True] bounds:  -0.9745025060367976 1.1873651355086268
maxiters 2
mask:  [ True False False  True  True  True  True] bounds:  -1.0136837226997282 -0.13492825427154725
maxiters 3
mask:  [ True False False  True  True  True  True] bounds:  -0.9558688522944143 -0.5743059884856376
```
![image](https://user-images.githubusercontent.com/17214293/88004390-e72cb100-cb41-11ea-9229-9ccc48841301.png)

**Comment by jehturner** [comment]
Thanks. As it happens I'm just looking at this stuff again and might add another option.

**Comment by saimn** [comment]
I don't really understand what means a "non-cumulative rejection". I mean, how is it different from doing no iterations, or how is it going to converge from one iteration to another if you keep using the points that were rejected at a previous iteration ?

In the example above, what should be the correct answer ? With 7 points and a 1-sigma threshold I don't really get the point.

**Comment by jehturner** [comment]
I can't remember exactly what I did to compare different masking methods in `FittingWithOutlierRemoval` before creating this issue, @saimn, but I'm thinking one would do the fitting and sigma estimation using the mask from the previous iteration, then re-calculate which points to reject, instead of accumulating newly-rejected points into the existing mask. What you say about it not converging from one iteration to another is essentially the problem I mentioned last week with our spline fitting in DRAGONS, which is not using the previous iteration's mask to calculate sigma...

Presumably the above answers should be about -0.33 or +0.29, respectively, which is the average of the 7 points being fitted (if I understand correctly)?  Unmasking the extra point actually would not help here, pushing the result further in the wrong direction in each case -- though it does illustrate that points rejected at one iteration might not remain so. Beyond that, I agree that this is not a great example of why non-cumulative rejection is needed; arguably, it demonstrates that a sigma-clipped median can converge in a biased way that cumulative masking might actually help constrain. Using a mean instead, the second example converges more sensibly to the middle point, but without demonstrating the point @ysBach was trying to make. Anyway, I don't know that this undermines the general point; maybe we can come up with a more realistic example, which would be useful for the docs if we do add the non-cumulative option.

**Comment by saimn** [comment]
Yes, I was mostly reacting to the example, which does not involve fitting and where the 1-sigma clipping is unlikely to give a good estimation of the mean. For the fitting case, I can see how accumulating the masked values or not can make a difference, since this will influence the mean estimator through the fit, but still I'm not sure how it can converge to a better result if you keep using the points that were rejected at the previous iterations. This is a kind of all or none problem, and there is no guarantee that one or the other solution will use the best set of points. The cumulative version seems the most sensible option in a general case, but that depends on other factors like the sigma threshold, and I guess using a robust std estimator would be better.
I guess there must be a better solution to the outliers rejection problem, maybe using weights ([1](https://en.wikipedia.org/wiki/Iteratively_reweighted_least_squares)), or in the case of scipy's least_squares using another loss function (soft_l1 or huber, [2](https://scipy-cookbook.readthedocs.io/items/robust_regression.html), [3](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.least_squares.html#scipy.optimize.least_squares)).

**Comment by jehturner** [comment]
This seems to me like a better example, where some (red) points that lie right on the fit after 3 iterations remain rejected:
```python
import numpy as np
from matplotlib import pyplot as plt

from astropy.stats import sigma_clip
from astropy.modeling.models import Gaussian1D, Chebyshev1D
from astropy.modeling.fitting import FittingWithOutlierRemoval, LinearLSQFitter

# Constant background with a spike (eg. cosmic ray) to be ignored:
pix = np.arange(100)
spike = Gaussian1D(amplitude=100., mean=80., stddev=1.5)
data = 10. + spike(pix)

# Add pseudo-random, shot-like noise:
np.random.seed(12345)
data += np.random.normal(loc=0., scale=1., size=data.shape[0]) * np.sqrt(data)

fitter = FittingWithOutlierRemoval(
    fitter=LinearLSQFitter(), outlier_func=sigma_clip, niter=3,
    sigma_lower=2, sigma_upper=2, maxiters=1
)

model, mask = fitter(Chebyshev1D(15), pix, data)
fit_vals = model(pix)

plt.plot(pix, data, '.', color='blue')
plt.plot(pix[mask], data[mask], '.', color='red')
plt.plot(pix, fit_vals, color='orange')
plt.axis((0, pix.shape[0], 0, 120.))
plt.show()
```
![cumrej_3](https://user-images.githubusercontent.com/322026/88219155-2b49af80-cc2f-11ea-8d76-2828c001c3b0.png)

**Comment by jehturner** [comment]
To allow comparison with non-cumulative rejection, I should point out that the final fit shown above is calculated *after* the final rejection, which is determined with respect to the fit from the previous iteration. Therefore red points lying close to the fit above should get restored by the next (4th) iteration, not necessarily at this one. But of course that won't happen since the rejection is cumulative and it turns out that the plot from 4 iterations is almost identical, just with a couple more points rejected towards the left:
![cumrej_4](https://user-images.githubusercontent.com/322026/88248006-b053b980-cc6d-11ea-9300-dd76cd99cee3.png)

**Comment by jehturner** [comment]
Now if I quickly hack this code path to make rejection non-cumulative, this is what I get after 4 iterations:
![newrej_4](https://user-images.githubusercontent.com/322026/88249050-31f91680-cc71-11ea-99a3-22f92fdf98f8.png)
Here I've made `SigmaClip._sigmaclip_noaxis` discard its input mask after calculating rejection thresholds and produce a new one from those limits. I also had to avoid the arithmetic with masked arrays in `FittingWithOutlierRemoval`, which produces unexpected results because it leaves masked values unchanged. That's enough to illustrate the difference, but it's just a hack and more work (and probably thought about the intended behaviour) would be needed for production.

**Comment by jehturner** [comment]
Of course, this begs the question of how the fit looks after 5 iterations, which is where unmasking those no-longer-deviant points would make a difference. The answer is "almost identical to 4 iterations", so whether it's worth worrying about is questionable:
![newrej_5](https://user-images.githubusercontent.com/322026/88250830-c87c0680-cc76-11ea-9f37-9b8fce06a7ac.png)
I haven't actually identified a use case where non-cumulative masking is really helpful yet...

**Comment by mcara** [comment]
Default `niter` is 3 and you are iterating more than that. Could you, please, show your plots above for 1st, 2nd, and 3rd iterations? The fact that iterations converge fast alone do not prove that the non-cumulative method is not useful.

**Comment by saimn** [comment]
This fit seems quite sensitive to the number of iterations:
![image](https://user-images.githubusercontent.com/311639/88296354-93020800-cccc-11ea-952d-045d1e4404ba.png)
Probably because the polynomial degree is a bit high if the goal is to reject the line, and the sigma a bit low? Here is another fit with degree=10 and sigma=2.5:
![image](https://user-images.githubusercontent.com/311639/88296497-c3e23d00-cccc-11ea-8ba5-d3842a9577df.png)

There are various ways to solve this problem, and I guess on real data there would be the same trade-offs.  

But anyway, on astropy side I guess it's reasonable to provide both options, cumulative and non-cumulative, and give the choice to users.

**Comment by jehturner** [comment]
I don't claim to prove that the non-cumulative method is not useful -- but OTOH no-one has come up with a very good example where it is yet...
